File: examples/dcps/Durability/C++/src/DurabilityDataPublisher.cpp

/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to 2011 PrismTech 
 *   Limited and its licensees. All rights reserved. See file:
 *
 *                     $OSPL_HOME/LICENSE 
 *
 *   for full copyright notice and license terms. 
 *
 */

/************************************************************************
 * LOGICAL_NAME:    DurabilityDataPublisher.cs
 * FUNCTION:        OpenSplice Tutorial example code.
 * MODULE:          Tutorial for the C# programming language.
 * DATE             September 2010.
 ************************************************************************
 * 
 * This file contains the implementation of the Publisher for the 'Durability' example.
 * 
 ***/
#include <string>
#include <sstream>
#include <iostream>
#include "DDSEntityManager.h"
#include "ccpp_DurabilityData.h"
#include "os.h"

using namespace DDS;
using namespace DurabilityData;

void usage()
{
  //usage : DurabilityDataPublisher <durability_kind> <autodispose_flag>
  //		. durability_kind = transient | persistent
  //		. autodispose_flag = false | true
  cerr << "*** ERROR" << endl;
  cerr << "*** usage : DurabilityDataPublisher <durability_kind> <autodispose_flag> <automatic_flag>"
    << endl;
  cerr << "***         . durability_kind = transient | persistent" << endl;
  cerr << "***         . autodispose_flag = false | true" << endl;
  cerr << "***         . automatic_flag = false | true" << endl;
 exit( - 1);
}

int main(int argc, char *argv[])
{
  bool automatic = true;
  ReturnCode_t status =  - 1;
  os_time delay_180s = { 180, 0 };
  os_time delay_1s = { 1, 0 };

  if (argc < 4)
  {
    usage();
  }
  if ((strcmp(argv[1], "transient") && strcmp(argv[1], "persistent")) || 
    (strcmp(argv[2], "false") && strcmp(argv[2], "true")))
  {
    usage();
  }
  string durability_kind(argv[1]);
  bool autodispose_unregistered_instances = (strcmp(argv[2], "true") == 0);
  DDSEntityManager *mgr = new DDSEntityManager(durability_kind,
    autodispose_unregistered_instances);
  
  automatic = (strcmp(argv[3], "true") == 0);

  // Wait for the Subscriber (case of transient = true && auto_dispose = true)
  os_nanoSleep(delay_1s);

  // create domain participant
  char partition_name[] = "Durability example";
  mgr->createParticipant(partition_name);

  //create type
  MsgTypeSupport_var mt = new MsgTypeSupport();
  mgr->registerType(mt.in());

  //create Topic
  char topic_name[] = "DurabilityData_Msg";
  mgr->createTopic(topic_name);

  //create Publisher
  mgr->createPublisher();

  // create DataWriter
  mgr->createWriter();

  // Publish Events
  ostringstream buf;
  DataWriter_ptr dwriter = mgr->getWriter();
  MsgDataWriter_var DurabilityDataWriter = MsgDataWriter::_narrow(dwriter);

  Msg *instances[10];
  InstanceHandle_t userHandle[10];
  for (int x = 0; x < 10; x++)
  {
    instances[x] = new Msg();
    instances[x]->id = x;
    buf.str(string(""));
    buf << x;
    userHandle[x] = DurabilityDataWriter->register_instance(*instances[x]);
    instances[x]->content = CORBA::string_dup(buf.str().c_str());
    cout << instances[x]->content << endl;
    DurabilityDataWriter->write(*instances[x], userHandle[x]);

  }
  if (! automatic) {
     char c = 0;
     cout << "Enter E to exit" << endl;
     while (c != 'E')
     {
       cin >> c;
     }
  } 
  else {
    cout << "=== sleeping 180s..." << endl;
    os_nanoSleep(delay_180s);   
  }
  //clean up

  // transient and autodispose unregister instances
  if(autodispose_unregistered_instances)
  {
	  for (int x=0; x<10; x++)
	  {		  
		  status = DurabilityDataWriter->unregister_instance(*instances[x], userHandle[x]);
		  checkStatus(status, "DataWriter::unregister_instance");
	  }
  }

  /* Do not release the data-samples in
  order to let persistency test work*/

  /* Remove the DataWriters */
  mgr->deleteWriter(DurabilityDataWriter.in ());

  /* Remove the Publisher. */
  mgr->deletePublisher();

  /* Remove the Topics. */
  mgr->deleteTopic();

  /* Remove Participant. */
  mgr->deleteParticipant();

  return 0;
}

File: examples/dcps/Durability/java/src/DurabilityDataPublisher.java
/*
 *                         Vortex OpenSplice
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

import java.io.IOException;

import DurabilityData.*;
import DDS.*;

public class DurabilityDataPublisher {

	public static void usage() {
		System.err.println("*** ERROR");
		System.err
				.println("*** usage : DurabilityDataPublisher <durability_kind> <autodispose_flag> <automatic_flag>");
		System.err
				.println("***         . durability_kind = transient | persistent");
		System.err.println("***         . autodispose_flag = false | true");
		System.err.println("***         . automatic_flag = false | true");
	}

	public static void main(String args[]) {

		if (args.length < 3) {
			usage();
		}
		if ((!args[0].equals("transient") && !args[0].equals("persistent"))
				|| (!args[1].equals("false") && !args[1].equals("true"))) {
			usage();
		}
		String durability_kind = args[0];
		boolean automatic = true;
		boolean autodispose_unregistered_instances = (args[1].equals("true"));
		boolean isPersistent = (args[0].equals("persistent"));

		DDSEntityManager mgr = new DDSEntityManager(durability_kind,
				autodispose_unregistered_instances);
		automatic = (args[2].equals("true"));
		// create domain participant
		String partition_name = "Durability example";
		mgr.createParticipant(partition_name);

		// create type
		MsgTypeSupport mt = new MsgTypeSupport();
		mgr.registerType(mt);

		// create Topic
		String topic_name = "JavaDurabilityData_Msg";
		if (isPersistent) {
            topic_name = "PersistentJavaDurabilityData_Msg";
        }
		mgr.createTopic(topic_name);

		// create Publisher
		mgr.createPublisher();

		// create DataWriter
		mgr.createWriter();

		// Publish Events
		DataWriter dwriter = mgr.getWriter();
		MsgDataWriter DurabilityDataWriter = MsgDataWriterHelper
				.narrow(dwriter);
		int status;

		Msg msgSample = new Msg();
		DurabilityDataWriter.register_instance(msgSample);
		for (int i = 0; i < 10; i++) {
			msgSample.id = i;
			msgSample.content = "" + i;
			status = DurabilityDataWriter.write(msgSample, HANDLE_NIL.value);
			ErrorHandler.checkStatus(status, "MsgDataWriter.write");

		}

		if (!automatic) {
			char c = 0;
			System.out.println("Enter E to exit");
			while (c != 'E') {
				try {
					c = (char) System.in.read();
					System.in.skip(System.in.available());
					System.out.println(c);
				} catch (IOException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
			}
		} else {
			try {
				Thread.currentThread().sleep(30000);// sleep for 30 s
			} catch (InterruptedException ie) {
			}
		}
		// clean up

		/*
		 * Do not release the data-samples in order to let persistency test work
		 */

		/* Remove the DataWriters */
		mgr.getPublisher().delete_datawriter(DurabilityDataWriter);

		/* Remove the Publisher. */
		mgr.deletePublisher();

		/* Remove the Topics. */
		mgr.deleteTopic();

		/* Remove Participant. */
		mgr.deleteParticipant();

	}
}
